constellation/constellation.py

```diff
@@ -0,0 +1,236 @@
+                else:
+                    container.stop()
+
+    def remove(self, prefix):
+        container = self.get(prefix)
+        if container:
+            print("Removing '{}'".format(self.name))
+            with docker_util.ignoring_missing():
+                container.remove()
+
+
+class ConstellationContainerCollection:
+    def __init__(self, collection):
+        self.collection = collection
+
+    def get(self, name, prefix):
+        for x in self.collection:
+            if x.name == name:
+                return x.get(prefix)
+        raise Exception("Container '{}' not defined".format(name))
+
+    def exists(self, prefix):
+        return [x.exists(prefix) for x in self.collection]
+
+    def _apply(self, method, *args):
+        for x in self.collection:
+            x.__getattribute__(method)(*args)
+
+    def pull_images(self):
+        self._apply("pull_image")
+
+    def stop(self, prefix, kill=False):
+        self._apply("stop", prefix, kill)
+
+    def remove(self, prefix):
+        self._apply("remove", prefix)
+
+    def start(self, prefix, network, volumes, data=None):
+        self._apply("start", prefix, network, volumes, data)
+
+
+class ConstellationVolume:
+    def __init__(self, role, name):
+        self.role = role
+        self.name = name
+
+    def exists(self):
+        return docker_util.volume_exists(self.name)
+
+    def create(self):
+        docker_util.ensure_volume(self.name)
+
+    def remove(self):
+        docker_util.remove_volume(self.name)
+
+
+class ConstellationVolumeCollection:
+    def __init__(self, volumes):
+        if not volumes:
+            self.collection = []
+        else:
+            self.collection = [ConstellationVolume(k, v) for k, v in
+                               volumes.items()]
+
+    def get(self, role):
+        for x in self.collection:
+            if x.role == role:
+                return x.name
+        raise Exception("Mount with role '{}' not defined".format(role))
+
+    def create(self):
+        for vol in self.collection:
+            vol.create()
+
+    def remove(self):
+        for vol in self.collection:
+            vol.remove()
+
+
+class ConstellationNetwork:
+    def __init__(self, name):
+        self.name = name
+
+    def exists(self):
+        return docker_util.network_exists(self.name)
+
+    def create(self):
+        docker_util.ensure_network(self.name)
+
+    def remove(self):
+        docker_util.remove_network(self.name)
+
+
+class ConstellationMount:
+    def __init__(self, name, path, **kwargs):
+        self.name = name
+        self.path = path
+        self.kwargs = kwargs
+
+    def to_mount(self, volumes):
+        return docker.types.Mount(self.path, volumes.get(self.name),
+                                  **self.kwargs)
+
+
+# only handles the simple case of "expose a port" and not "remap a
+# port", and assumes the port is to be exposed onto all interfaces.
+def container_ports(ports):
```

Comment:
I feel like we may want to do other things with ports in the not too distant future (e.g. in Montagu we map the annex from 5432 to 5433)

Reply:
yes, for sure. I guess we can accept something like `[{from: to}]` so that `[x]` is a shorthand for `[x: x]. Logged this as mrc-659